libtenzir/src/data_builder.cpp

```diff
@@ -0,0 +1,1233 @@
+            default:
+              TENZIR_UNREACHABLE();
+          }
+        } else if (seed_idx
+                   == caf::detail::tl_index_of<field_type_list,
+                                               duration>::value) {
+          auto unit = seed->attribute("unit").value_or("s");
+          if constexpr (std::same_as<T, int64_t>) {
+            auto res = cast_value(int64_type{}, v, duration_type{}, unit);
+            if (res) {
+              data(*res);
+              return;
+            }
+          } else if constexpr (std::same_as<T, uint64_t>) {
+            auto res = cast_value(uint64_type{}, v, duration_type{}, unit);
+            if (res) {
+              data(*res);
+              return;
+            }
+
+          } else if constexpr (std::same_as<T, double>) {
+            auto res = cast_value(double_type{}, v, duration_type{}, unit);
+            if (res) {
+              data(*res);
+              return;
+            }
+          }
+        } else if (seed_idx
+                   == caf::detail::tl_index_of<field_type_list, time>::value) {
+          auto unit = seed->attribute("unit");
+          if (not unit) {
+            rb.emit_or_throw(
+              diagnostic::warning("could not parse value as `{}`", time_type{})
+                .note("the read value as a number, but the schema does not "
+                      "specify a unit"));
+            return;
+          }
+          if constexpr (std::same_as<T, int64_t>) {
+            auto res = cast_value(int64_type{}, v, duration_type{}, *unit);
+            if (res) {
+              data(time{} + *res);
+              return;
+            }
+          } else if constexpr (std::same_as<T, uint64_t>) {
+            auto res = cast_value(uint64_type{}, v, duration_type{}, *unit);
+            if (res) {
+              data(time{} + *res);
+              return;
+            }
+
+          } else if constexpr (std::same_as<T, double>) {
+            auto res = cast_value(double_type{}, v, duration_type{}, *unit);
+            if (res) {
+              data(time{} + *res);
+              data(*res);
+              return;
+            }
+          }
+        }
+      } else if (seed_idx == type_index_string) {
+        // stringify if possible
+        if constexpr (fmt::is_formattable<T>{}) {
+          data(fmt::format("{}", v));
+          return;
+        }
+      }
+      // TODO this happens in our intentionally "broken" zeek.json event in the
+      // test input, where we have field `id : 0`, whereas the schema expects
+      // `id : zeek.conn_id` The resulting issue is that a protected/preparsed
+      // series_builder will reject the value entirely This could be resolved by
+      // not preparing any builders in the `multi_series_builder` and instead
+      // ensuring that all fields top level are written (as null) on commit.
+      // This is a non-trivial effort though and should be considered as a
+      // follow-up to precise parsing.
```

Comment:
If we are writing into a prepared `series_builder` and have mismatch that cannot be resolved, the field will be null instead of its actual value in the event.

The only way to resolve this and keep the actual events fields would be to not prepare any builders. This would affect the order of the fields (fields from the schema would no longer come first).

Reply:
Okay, I see. I didn't understand purely from the comment what "problem" this was referring to. Now I do.